ceviche/fdfd.py

```diff
@@ -191,6 +191,123 @@ def _solve_fn(self, eps_vec, entries_a, indices_a, Jz_vec):
+        Nfreq = npa.shape(self.delta)[0]
+        delta_matrix = self.delta.reshape([Nfreq, npa.prod(self.shape)])
+        phi_matrix = self.phi.reshape([Nfreq, npa.prod(self.shape)])
+        # create the A matrix for this polarization
+        entries_a, indices_a = self._make_A(eps_vec, delta_matrix, phi_matrix)
+
+        # solve field componets usng A and the source
+        Fx_vec, Fy_vec, Fz_vec = self._solve_fn(eps_vec, entries_a, indices_a, source_vec)
+
+        # put all field components into a tuple, convert to grid shape and return them all
+        Fx = self._vec_to_grid(Fx_vec)
+        Fy = self._vec_to_grid(Fy_vec)
+        Fz = self._vec_to_grid(Fz_vec)
+
+        return Fx, Fy, Fz
+
+    def _make_A(self, eps_vec, delta_matrix, phi_matrix):
+
+        M = 2*self.Nsb + 1
+        N = self.Nx * self.Ny
+        W = self.omega + npa.arange(-self.Nsb,self.Nsb+1)*self.Omega
+
+        C = sp.kron(sp.eye(M), - 1 / MU_0 * self.Dxf.dot(self.Dxb) - 1 / MU_0 * self.Dyf.dot(self.Dyb))
+        entries_c, indices_c = get_entries_indices(C)
+
+        # indices into the diagonal of a sparse matrix 
+        D = - EPSILON_0 * sp.kron(sp.spdiags(W**2,[0],M,M), sp.spdiags(eps_vec,[0],N,N))
+        entries_diag, indices_diag = get_entries_indices(D)
+
+        entries_a = npa.hstack((entries_diag, entries_c))
+        indices_a = npa.hstack((indices_diag, indices_c))
+
+        Nfreq = npa.shape(delta_matrix)[0]
+        for k in npa.arange(Nfreq):
+            delta_diag_p = - EPSILON_0 * sp.spdiags(delta_matrix[k,:] * npa.exp(1j*phi_matrix[k,:]), [0], N, N)
+            delta_diag_m = - EPSILON_0 * sp.spdiags(delta_matrix[k,:] * npa.exp(-1j*phi_matrix[k,:]), [0], N, N)
+            Z = sp.kron(sp.spdiags(0.5*npa.concatenate([ [0], W[:-(k+1)] ])**2,[k+1],M,M), delta_diag_p)
+            entries_z, indices_z = get_entries_indices(Z)
+            entries_a = npa.hstack((entries_z, entries_a))
+            indices_a = npa.hstack((indices_z,indices_a))
+            Z = sp.kron(sp.spdiags(0.5*npa.concatenate([ W[k+1:], [0] ])**2,[-(k+1)],M,M), delta_diag_m)
+            entries_z, indices_z = get_entries_indices(Z)
+            entries_a = npa.hstack((entries_z, entries_a))
+            indices_a = npa.hstack((indices_z,indices_a))
+
+        return entries_a, indices_a
+
+    def _solve_fn(self, eps_vec, entries_a, indices_a, Jz_vec):
+        M = 2*self.Nsb + 1
+        N = self.Nx * self.Ny
+        W = self.omega + npa.arange(-self.Nsb,self.Nsb+1)*self.Omega 
+        P = sp.kron(sp.spdiags(W,[0],M,M), sp.eye(N))
+        entries_p, indices_p = get_entries_indices(P)
+        b_vec = 1j * sp_mult(entries_p,indices_p,Jz_vec)
+        Ez_vec = sp_solve(entries_a, indices_a, b_vec)
+        Hx_vec, Hy_vec = self._Ez_to_Hx_Hy(Ez_vec)
+        return Hx_vec, Hy_vec, Ez_vec
+
+    def _Ez_to_Hx(self, Ez_vec):
+        M = 2*self.Nsb + 1
+        Winv = npa.reciprocal(self.omega + npa.arange(-self.Nsb,self.Nsb+1)*self.Omega)
```

Comment:
maybe 
`Winv = 1 / (self.omega + npa.arange(-self.Nsb, self.Nsb+1) * self.Omega)`
is a bit easier to understand

Reply:
Modified.